Fix rubocop offences

isubas opened the pull request:
#### Bu PR'in yaptığı işi/değişikliği ve bu işi/değişikliği neden yaptığını açıklayın

Yeni eklenen cop'lar sonucu oluşan offence'leri düzeltir.

`Performance/ChainArrayAllocation` için tüm düzeltmeleri yapmadım, değişmesi gereken yerler var bu yüzden zaman içerisinde düzelterek gideceğiz.

#### İlgili/kapatılacak iş kayıtları

Fix #1856 

#### Teknik borç kayıtları

N/A

#### Veritabanına etkileri

N/A

#### Sistem etkileri

N/A

#### Kontrol listesi

- [X] [Katkı sağlama dokümanını](../blob/master/.github/CONTRIBUTING.md) okudum
- [X] İş kaydının başlığı kurallara (sadece ilk harf büyük, emir kipinde problem cümlesi) uygun
- [ ] Yapılan iş/değişikliği dokümante ettim
- [ ] Yapılan iş/değişikliğin testlerini yazdım
- [X] Test coverage oranını kontrol ettim
- [X] Kod kalitesi (karma) ve test suite dahil olmak üzere tüm entegre kontroller başarıyla geçiyor
- [X] Kendimi bu PR'e assign ettim
- [X] Yapılan iş/değişiklik ile ilgili proje üyelerinden review talep ettim
- [X] Gerekli etiketlemeyi (ör. bug) yaptım

#### Ek içerik

[//]: # (Kaynaklar)
[//]: # (Dış bağlantılar)
[//]: # (Ekran görüntüleri)
[//]: # (Örnek çözümler)

isubas committed:
Fix rubocop offences

isubas committed:
Fix erb linter offences

isubas committed:
Minor change

isubas committed:
Remove --autocorrect flag from erb linter task

isubas committed:
Gelecekte tekrardan gözden geçirmek üzere ErbSafety kuralını disable et

isubas commented:
erb_linter'da `ErbSafety`'yi şimdilik disable ettim verdiği uyarıların bir çoğu gereksiz bunu geniş zamanda tekrardan kontrol edip son kararı veririm. Şimdilik böyle devam edelim.

ecylmz commented:
Kafayı yiyeceğim, `NewCops: enable` çalışmıyor. Travis'te de çalışmamış hala uyarılar var.
Eğer bunu komut satırı parametresi olarak verirsek çalışıyor. şöyle:

```bundle exec rails rubocop --pending-cops-enabled```

@isubas yerelde yukarıda yazdığım parametreyi vermeden rubocop'u çalıştırdığında aşağıdaki uyarıları görüyor musun? 
```
The following cops were added to RuboCop, but are not configured. Please set Enabled to either `true` or `false` in your `.rubocop.yml` file:
 - Layout/EmptyLinesAroundAttributeAccessor (0.83)
 - Layout/SpaceAroundMethodCallOperator (0.82)
 - Lint/DeprecatedOpenSSLConstant (0.84)
 - Lint/MixedRegexpCaptureTypes (0.85)
 - Lint/RaiseException (0.81)
 - Lint/StructNewOverride (0.81)
 - Style/ExponentialNotation (0.82)
 - Style/HashEachMethods (0.80)
 - Style/HashTransformKeys (0.80)
 - Style/HashTransformValues (0.80)
 - Style/RedundantFetchBlock (0.86)
 - Style/RedundantRegexpCharacterClass (0.85)
 - Style/RedundantRegexpEscape (0.85)
 - Style/SlicingWithRange (0.83)
For more information: https://docs.rubocop.org/rubocop/versioning.html
```

isubas commented:
Evet görüyorum rubocop'un bir bug'ı var gibi

dilara commented on app/services/ldap/entity.rb:
```diff
@@ -59,8 +58,8 @@ def preferredLanguage
             .map { |role| prefixes_for_role(role) }
-            .flatten
-            .uniq
+            .flatten!
```
> Performance/ChainArrayAllocation: Use unchained flatten! and uniq! (followed by return array if required) instead of chaining flatten...uniq.

uyarıda parantez içinde belirtilen durum sanırım burada geçerli. Yalnızca `student` ve yalnızca `employee` rolü olan iki kullanıcı için deneme yaptığımda `eduPersonAffiliation`, `nil` dönüyor.

isubas replied:
Bunun testini yazmış, normalde test geçiyor ama değin gibi konsol üzerinden denediğimde nil dönüyor

isubas commented:
@ecylmz NewCops: enable doğru çalışmıyor. Senin dediğin gibi çalıştırdığımda sorun var.

ecylmz commented:
Evet, `NewCops`'ta bir sıkıntı var. Şu tuhaf, `NewCops: foo` dersek foo diye bir seçenek yok diyor. `enable` dediğimizde bir şey demiyor ama çalışmıyor da. Ben anlamadım bir şey.

Bu soruna çözüm bulana kadar `bundle exec rails rubocop --pending-cops-enabled` ile durumu pending'te olanlar enable olduğunda ne gibi hatalarımız var onlara bakabiliriz.

Sorunu araştırmaya devam ediyorum.

isubas commented:
İnan usandım bu robucop'tan amma sorun çıkardı.

roktas commented:
@ecylmz Bu sorunu sevabına raporlayalım.

roktas commented:
Usanmayın, şükredin böyle bir araç var. :-)

isubas committed:
Rubocop Performance/ChainArrayAllocation kuralını bazı dosyalardan kaldır

roktas commented:
@isubas @ecylmz ve @huseyin'e assign ediyorum PR'ı, senin yeterince işin var.

roktas commented:
@huseyin @ecylmz özellikle ` Performance/ChainArrayAllocation` fixini yaparken dikkatli olun.  Eşdeğer kodu irb'de deneyerek ilerleyin bug olmasın.  İlgili doküman: https://rubocop.readthedocs.io/projects/performance/en/stable/cops_performance/#performancechainarrayallocation

isubas commented:
`Performance/ChainArrayAllocation` için irb denemeleri doğru sonuç vermiyor bilgilerinize, ben denedim.

isubas commented:
Hatta, dilara'nın söylediği durum test'lerden geçiyor ama hata veriyor. `LDAP::Entity` için `eduPersonAffiliation` test'i var ve ben bunları değiştirdikten sonra test geçti ama `nil` problemi oluşmuş. Dikkatli olmak şart.

roktas commented:
Bu bang'li versiyonla, bangsiz versiyon arasındaki semantik farktan kaynaklanıyor.  Örneğin `flatten` vs `flatten!` farkına resmi Ruby dokümanlarından bakmak lazım: https://ruby-doc.org/core-2.7.0/Array.html#method-i-flatten

![image](https://user-images.githubusercontent.com/47842/86485837-73795e80-bd62-11ea-830d-58a084188293.png)

![image](https://user-images.githubusercontent.com/47842/86485846-7a07d600-bd62-11ea-8a74-e59b851824a3.png)

Bang'li versiyon nil dönebiliyor.  Neden?  Semantiği böyle olduğundan "Returns nil if no modifications were made (i.e., the array contains no subarrays.)"

Sadece irb yetmez, dokümanlara da bakmak lazım.

ecylmz commented:
Yapılacak 3 iş var:

- Rubocop'a `NewCops: enable` özelliğinin çalışmadığını anlatan bir issue açmak. Bu bende.
- Rubocop'tan yanıt gelene kadar `bundle exec rails rubocop --pending-cops-enabled` komutunu elle çalıştırarak gelen `offenses`'leri çözmek. @huseyin buna sen bakabilir misin?
- Bunlar bitince, daha sonra bakalım dediğimiz kısımlar için `.rubocop_todo.yml` dosyası üreteceğiz.

isubas commented:
Ben'de proje açık ben hüseyin yerine bunları hızlıca yaparım. Bundan sonra oluşacaklara @huseyin bakar.

isubas committed:
rubocop_todo.yml dosyasını oluştur

isubas committed:
rubocop task'ine --enable-pending-cops opsiyonunu ekle

isubas committed:
Fix rubocop offencec

ecylmz commented:
Sorunu dün gece rubocop'a bildirdim: https://github.com/rubocop-hq/rubocop/issues/8234

Bugün bekleyelim, cevap gelmezse `--pending-cops-enabled` parametresiyle devam ederiz.

ecylmz committed:
Plugin'lerdeki .rubocop.yml dosyalarını kaldır

ecylmz committed:
Bazı autocorrect'leri geri al ve plugin'lerdeki sorunları todo'ya taşı

ecylmz committed:
erblinter için require ekle

roktas committed:
Plugin isimlendirmeleri Naming/FileName offense üretmesin

Bu bir sınır durum.  Plugin'leri Rails Engine mimarisine uygun olarak
adeta public bir gem yayınlayacakmış gibi kurguladık.  Bu durumda gem
adlarında başa nokul- ön eki geliyor.

Bu vesileyle bir rubocop exclude ilkesi koyalım.  Bu örnekte olduğu gibi
bazı ofansları (en azından orta vadede) TODO olarak kabul etmiyor ve
bilnçli olarak istisna koyuyoruz.  Bu yapıldığında aşağıdaki
aksiyonlardan biri alınmalı:

- Kural istinasını `.rubocop_todo.yml` dosyasından `.rubocop.yml`
  dosyasına al.  Özellikle grup halinde istisnalarda bunu tercih
  ediyoruz.

- İstisna çok bilinçli yapılmışsa ("ne yaptığımı biliyorum" durumu)
  istisnayı ilgili dosyada en kısa kapsamda `rubocop:disable` direktifiyle
  yerinde etkisiz hale getiriyoruz.  (Dikkat!  Bu tür kararlar
  review'lerde sorgulanabilir ve sorgulanmalıdır.  Bazen geliştiricinin
  niyeti anlaşılamayabiliyor.)

Bu komitte ik aksiyona başvuruldu.

roktas committed:
Tenant API için Rails/Delegate ofansını kalıcı olarak kaldır

Bir önceki komitte ifade edilen ikinci aksiyonu uyguladık ve istisnayı
dosyaya gömdük.

Neden?  Tenant API çeşitli meta programlama numaralarının döndüğü bir
yer ve burası kritik bir "code path".  Meta programlama üstüne bir de
ekstra Rails delegate meta bindirmek istemiyorum (bu olağan bir
delegasyon değil).  Cop'a uyup delegate kullandığımda şu an
hatırlamadığım bir sorun oldu.  Zorlarsam nedenini bulur ve çözerim.
Ama bunun önemi yok çünkü bu kritik kodda gelecekte yine patlama
olabileceğine işaret.  Özetle "attığımız taş ürküttüğümüz kurbağaya
değmeyecek".

roktas committed:
Rails Engine'in ürettiği dummy test uygulaması ofans üretmemeli

roktas committed:
Style/FormatStringToken ofansını düzelt

Style/FormatStringToken kuralı için Rubocop'ta "template" biçimini
seçmişiz, default "annotated" idi.  Nedeni şuydu: Default "annotated"
biçiminde `%<name>s` gibi sonda bir `s` bulunuyor, "template"de ise
basitçe `%{name}` diyebiliyoruz.

Fakat bu düzeltmelerde olduğu gibi uzunluk işin içine girdiğinde
`%<number>4d` yerine `%4{number}` kullanmak gerekiyor ve format stringte
"bu numeriktir" bilgisi (sondaki `d`) yok.  Bu bilginin kaybolmasının
herhangi bir yan etkisini görmedim ama bu durum default'un neden
"annotated" olduğunu da açıklıyor.  Kuralı (şimdilik) default'a çekelim
demiyorum ama gelecekte belki böyle yapabiliriz.  Biraz daha incelemem
lazım.

roktas committed:
Style/SingleLineBlockParams ofansını düzelt

roktas committed:
Style/ImplicitRuntimeError ofansını düzelt

roktas committed:
Metrics/ModuleLength, Naming/MethodName ofanslarını düzelt

LDAP entegrasyonundaki isimlendirmeler kendine özgü şartlara sahip, bunu
TODO'dan çıkartıp inline yaptık.

roktas committed:
Merge branch 'rubocop' of https://github.com/omu/nokul into rubocop

roktas committed:
Style/Send ofansını düzelt

roktas committed:
Gereksiz Rubocop istisnasını kaldır

roktas committed:
Minitest validation_helper ofanslarını kaldır

ecylmz commented:
Travis yeşillenirse içeri alacağım.

roktas committed:
Çok fena
